crates/ursa-tracker/src/main.rs

```diff
@@ -0,0 +1,222 @@
+use crate::{
+    ip_api::get_ip_info,
+    types::{Node, TrackerRegistration, PrometheusDiscoveryChunk},
+};
+use axum::http::StatusCode;
+use axum::{
+    extract::{ConnectInfo, Extension},
+    routing::{get, post},
+    Json, Router,
+};
+use hyper::HeaderMap;
+use serde_json::{json, Value};
+use rocksdb::{DB, IteratorMode, WriteBatch};
+use std::{env, net::SocketAddr, sync::Arc};
+use tracing::{info, error};
+use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
+
+mod ip_api;
+mod types;
+
+#[tokio::main]
+async fn main() {
+    tracing_subscriber::registry()
+        .with(tracing_subscriber::EnvFilter::new(
+            std::env::var("RUST_LOG").unwrap_or_else(|_| "info".into()),
+        ))
+        .with(tracing_subscriber::fmt::layer())
+        .init();
+
+    let db = Arc::new(DB::open_default("tracker_db").unwrap());
+    let token = env::var("IPINFO_TOKEN").expect("IPINFO_TOKEN is not set");
+
+    let app = Router::new()
+        .route("/register", post(registration_handler))
+        .route("/http_sd", get(http_sd_handler))
+        .layer(Extension(db))
+        .layer(Extension(token));
+
+    let addr = SocketAddr::from(([0, 0, 0, 0], 4000));
+    println!("Ursa-tracker listening on {}", addr);
+
+    axum::Server::bind(&addr)
+        .serve(app.into_make_service_with_connect_info::<SocketAddr>())
+        .await
+        .unwrap();
+}
+
+/// Track a new peer registration in the database
+async fn registration_handler(
+    headers: HeaderMap,
+    ConnectInfo(req_addr): ConnectInfo<SocketAddr>,
+    db: Extension<Arc<DB>>,
+    token: Extension<String>,
+    Json(registration): Json<TrackerRegistration>,
+) -> (StatusCode, Json<Value>) {
+    let id = registration.id;
+    info!("Received registration for: {}", id);
+
+    // todo: registration verification
+
+    let addr = registration.addr.clone().unwrap_or_else(|| {
+        // if no dns or ip address is provided, use the address of the request.
+        // Prefer X-Forwarded-For header if present from reverse proxy. otherwise, use the
+        // address of the request
+        let ip = headers
+            .get("X-Forwarded-For")
+            .map(|x| x.to_str().unwrap().to_string())
+            .unwrap_or_else(|| req_addr.ip().to_string());
+        (ip != "127.0.0.1").then_some(ip).unwrap_or_default()
```

Comment:
this code possibly allows an attacker to cause the termination of the server by putting non-visible ASCII characters in the `X-Forwarded-For` header in scenarios where there is no reverse-proxy overwrite of the said header.

this is caused by the `x.to_str().unwrap()` part of the code.

Reply:
Good catch! Sneaky unwrap. Though it's low priority as we are the ones running this tracker and it's setup through docker and proxies through nginx, which will always overwrite this header